### PR TITLE
Lab 24 Kaylyn

### PR DESCRIPTION
**Pull request by kaylynyuh**

### CODE REVIEW

File: lab-kaylyn/app/component/game-history/_game-history.scss [hunk 1]
```diff
@@ -0,0 +1,40 @@
+@import "theme";
+
+.game-history {
+  margin: 2%;
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/_game-history.scss]: @kaylynyuh I see a few `2%` and `5%` margins - may be best to set these up as `_vars` so that you can adjust app wide if need be...  for example:
`$gutter-lg: 5%` `$gutter-md: $gutter-lg / 2` `$gutter-sm: 1%`, etc etc.

File: lab-kaylyn/app/component/game-history/_game-history.scss [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    li {
+      margin-top: 5%;
+      margin-bottom: 2%;
+      padding: 5px;
+      border-radius: 5px;
+      color: $app-white;
+      background-color: $app-2nd;
+
+      p {
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/_game-history.scss]: @kaylynyuh good use of nesting

File: lab-kaylyn/app/component/game-history/_game-history.scss [hunk 3]
```diff
@@ -0,0 +1,40 @@
+@import "theme";
+
+.game-history {
+  margin: 2%;
+
+  ol {
+    padding: 0px;
+    list-style: none;
+    letter-spacing: .075rem;
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/_game-history.scss]: @kaylynyuh good use of `rem` and `em`

File: lab-kaylyn/app/component/game-history/_game-history.scss [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      }
+    }
+  }
+}
+
+.form-control {
+  margin-top: 3em;
+  margin-left: .5em;
+  width: 25%;
+  display: inline;
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/_game-history.scss]: @kaylynyuh for style points, you can order your properties in this fashion (this is an opinion and not a rule):

```
position, display, margin, padding, color, text
```

File: lab-kaylyn/app/component/game-history/game-history.html [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<section class="game-history">
+  <ol>
+    <li ng-repeat="turn in gameHistoryCtrl.history">
+      <h3>turn {{turn.turn}} : {{turn.location}}</h3>
+      <p>HealthPoints: {{turn.healthPoints}}</p>
+      <p>{{turn.desc}}</p>
+    </li>
+  </ol>
+</section>
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/game-history.html]: @kaylynyuh game history markup looks good

File: lab-kaylyn/app/component/game-history/game-history.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  template: require('./game-history.html'),
+  controller: 'GameHistoryController',
+  controllerAs: 'gameHistoryCtrl',
+});
+
+ngAdventure.controller('GameHistoryController', ['$log', 'playerService', GameHistoryController]);
+function GameHistoryController($log, playerService){
+  $log.debug('init gameHistoryCtrl');
+  this.history = playerService.history;
+}
```
Comment (bnates) [inline lab-kaylyn/app/component/game-history/game-history.js]: @kaylynyuh game history component looks good

File: lab-kaylyn/app/component/game-pad/_game-pad.scss [hunk 7]
```diff
@@ -0,0 +1,5 @@
+@import "_theme";
```
Comment (bnates) [inline lab-kaylyn/app/component/game-pad/_game-pad.scss]: @kaylynyuh no need to include the underscore at the beginning of the file name when importing a partial - you can just do:
`@import "partial_name"`

File: lab-kaylyn/app/component/game-pad/game-pad.html [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  <form novalidate name="game-pad">
+    <label for="direction">Direction</label>
+    <select ng-model="gamePadCtrl.moveDirection" class="form-control">
+      <option ng-repeat="direction in gamePadCtrl.directions" value="{{direction}}">
+        {{direction}}
+      </option>
+    </select>
+    <button type="button" class="btn btn-default" ng-click="gamePadCtrl.movePlayer()">MOVE</button>
+  </form>
+</section>
```
Comment (bnates) [inline lab-kaylyn/app/component/game-pad/game-pad.html]: @kaylynyuh game pad markup looks good

File: lab-kaylyn/app/component/game-pad/game-pad.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  this.movePlayer = function(){
+    playerService.movePlayer(this.moveDirection)
+    .then(location => {
+      $log.log(`Arya's currently located in ${location}`);
+    })
+    .catch(err => {
+      $log.error(err);
+    });
+  };
+}
```
Comment (bnates) [inline lab-kaylyn/app/component/game-pad/game-pad.js]: @kaylynyuh game pad component looks good

File: lab-kaylyn/app/component/player-info/_player-info.scss [hunk 10]
```diff
@@ -0,0 +1,14 @@
+@import "theme";
+
+.player-info {
+  margin: 5px;
```
Comment (bnates) [inline lab-kaylyn/app/component/player-info/_player-info.scss]: @kaylynyuh got a bunch of `5px`'s - maybe make a `_var` for this?
ex: `$margin-sm`? - This should be separate from your gutters `_vars`, in the event that you are using pixels vs percentages (or other metric sizing units - `em`, `rem`, `vm`, `vh`)

File: lab-kaylyn/app/component/player-info/player-info.html [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<section class="player-info">
+  <h3>You are: {{playerInfoCtrl.player.name}}</h3>
+  <p>Location: {{playerInfoCtrl.player.location}}</p>
+  <p>Health Points: {{playerInfoCtrl.player.healthPoints}}</p>
+  <p>Navigate yourself through the 7 Kingdoms in search of Nymeria. Be weary of your health. Safe Travels and good luck.</p>
+</section>
```
Comment (bnates) [inline lab-kaylyn/app/component/player-info/player-info.html]: @kaylynyuh player markup looks good

File: lab-kaylyn/app/component/player-info/player-info.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  controller: 'PlayerInfoController',
+  controllerAs: 'playerInfoCtrl',
+});
+
+ngAdventure.controller('PlayerInfoController', ['$log', 'playerService', PlayerInfoController]);
+
+function PlayerInfoController($log, playerService){
+  $log.debug('init PlayerInfoController');
+  this.player = playerService.player;
+}
```
Comment (bnates) [inline lab-kaylyn/app/component/player-info/player-info.js]: @kaylynyuh player stats component looks good

File: lab-kaylyn/app/entry.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+//create angular module
+angular.module('ngAdventure', []);
+
+//require services
+require('./service/map-service.js');
+require('./service/player-service.js');
+//require components
+require('./component/game-pad/game-pad.js');
+require('./component/player-info/player-info.js');
+require('./component/game-history/game-history.js');
```
Comment (bnates) [inline lab-kaylyn/app/entry.js]: @kaylynyuh entry file looks good

File: lab-kaylyn/app/index.html [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    <title>Game Of Thrones</title>
+  </head>
+  <body>
+    <h1>Help Arya Stark Find Her Direwolf</h1>
+    <h4>Oh, and welcome to The Game of Thrones...</h4>
+    <player-info> </player-info>
+    <game-pad> </game-pad>
+    <game-history> </game-history>
+  </body>
+</html>
```
Comment (bnates) [inline lab-kaylyn/app/index.html]: @kaylynyuh index markup and associated components and directives look good!

File: lab-kaylyn/app/scss/lib/_custom_bootstrap.scss [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+// Components w/ JavaScript
+//@import "~bootstrap-sass/assets/stylesheets/bootstrap/modals";
+//@import "~bootstrap-sass/assets/stylesheets/bootstrap/tooltip";
+//@import "~bootstrap-sass/assets/stylesheets/bootstrap/popovers";
+//@import "~bootstrap-sass/assets/stylesheets/bootstrap/carousel";
+
+// Utility classes
+@import "~bootstrap-sass/assets/stylesheets/bootstrap/utilities";
+@import "~bootstrap-sass/assets/stylesheets/bootstrap/responsive-utilities";
```
Comment (bnates) [inline lab-kaylyn/app/scss/lib/_custom_bootstrap.scss]: @kaylynyuh i generally comment everything in this file out, with the exception of reset, mixin, and base styles - then simply uncomment any additional partials, as needed, during the dev process

File: lab-kaylyn/app/scss/lib/_theme.scss [hunk 16]
```diff
@@ -0,0 +1,8 @@
+$app-black : #000;
+$app-white : #fff;
+$app-1st: #3D312E;
```
Comment (bnates) [inline lab-kaylyn/app/scss/lib/_theme.scss]: @kaylynyuh using numbers in your `_var` names is fine - that said, my preference is to not do so - it looks a _bit_ more professional

File: lab-kaylyn/app/scss/main.scss [hunk 17]
```diff
@@ -0,0 +1,23 @@
+@import "_vendor";
+@import "_theme";
```
Comment (bnates) [inline lab-kaylyn/app/scss/main.scss]: @kaylynyuh just reiterating - no need for the prepended underscore here

File: lab-kaylyn/app/scss/main.scss [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+h1 {
+  text-align: center;
+  margin-top: 2em;
+  padding-top: 1.5em;
+}
+
+h4 {
+  text-align: center;
+  margin-bottom: 2em;
+}
```
Comment (bnates) [inline lab-kaylyn/app/scss/main.scss]: @kaylynyuh base styles are on point

File: lab-kaylyn/app/service/map-service.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+      south: 'Sunspear',
+      east: null,
+      west: 'Highgarden',
+    },
+    'Sunspear': {
+      desc: 'You are in Sunspear, House Martell',
+      north: 'The Iron Islands',
+      south: null,
+      east: null,
+      west: null,
```
Comment (bnates) [inline lab-kaylyn/app/service/map-service.js]: @kaylynyuh using `null` is fine _but_ this can be abstracted out so that you just check if a direction has a `null` value in your logic

File: lab-kaylyn/app/service/player-service.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+        healthPoints: player.healthPoints,
+        healthWarning: null,
+      });
+      console.log('history', history);
+      player.location = newLocation;
+      return resolve(player.location);
+    });
+  };
+  return service;
+}
```
Comment (bnates) [inline lab-kaylyn/app/service/player-service.js]: @kaylynyuh history service looks good

File: lab-kaylyn/karma.conf.js [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Continuous Integration mode
+    // if true, Karma captures browsers, runs the tests and exits
+    singleRun: false,
+
+    // Concurrency level
+    // how many browser should be started simultaneous
+    concurrency: Infinity,
+  });
+};
```
Comment (bnates) [inline lab-kaylyn/karma.conf.js]: @kaylynyuh karma config looks good - you can remove all of the comments if you wish - not required but does help with filesize and readability since the option names (property names) are self explanatory

File: lab-kaylyn/test/game-pad-ctrl-test.js [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+  beforeEach(() => {
+    angular.mock.module('ngAdventure');
+
+    angular.mock.inject($controller => {
+      this.gamePadCtrl = new $controller('GamePadController');
+    });
+  });
+});
+
+describe('testing initial properties', () => {
```
Comment (bnates) [inline lab-kaylyn/test/game-pad-ctrl-test.js]: @kaylynyuh basic tests but good tests - as you get better at frontend dev/testing, you'll start to pickup on edge cases and can account for those in your tests
